From 7ea427a78a80bcab7b924b86b86bc09101238bfc Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 08 Oct 2020 11:08:10 +0300
Subject: [PATCH] [SW-22078] ppboot: goya2: bring arc out of reset

Currently in BTL based flow, ARCs are not used for any operations.
There is requirement to run BTL based flow till preboot and then load
zephyr on ARC0 and ARC1 as part of boot fit flow. This patch add changes
to perform reset vector change and bring both ARCs out of reset as part
of ppboot flow. This is required to copy fit images in ARC ICCM and DCCM
memories.

Change-Id: Idac917aa5891246df6f19fd9abd0944f42da1c0c
---

diff --git a/goya2/common.c b/goya2/common.c
index 41a0d1e..72d7efc 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -15,6 +15,10 @@
 #include <pcie_dbi_regs.h>
 
 #include <cpu_timestamp_regs.h>
+#include <psoc_arc0_cfg_regs.h>
+#include <psoc_arc0_cfg_masks.h>
+#include <psoc_arc1_cfg_regs.h>
+#include <psoc_arc1_cfg_masks.h>
 
 #include <hl_boot_if.h>
 #include <goya2/goya2_reg_map.h>
@@ -211,7 +215,25 @@
 		;
 }
 
-void hl_plat_init(void) {}
+static void init_arc(void)
+{
+	/*
+	 * Program Reset vector address for ARC0 and ARC1 to ICCM 0x0
+	 * location where vector table first entry will be available and
+	 * then bring ARC0 and ARC1 out of reset. ARCs when out of reset
+	 * is by default in halt state.
+	 */
+	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
+	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
+	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC1_CFG_RST_VEC_ADDR);
+	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC1_CFG_ARC_RST);
+
+}
+
+void hl_plat_init(void)
+{
+	init_arc();
+}
 
 void ppboot_log_waiting_for_pcie(void)
 {
diff --git a/goya2/common.h b/goya2/common.h
index 8dd1ae8..0400687 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -17,3 +17,6 @@
 #define FUSE_WORDS_PER_BANK		0x40
 #define I2C_SLAVE_ADDRESS		0x6A
 #define I2C_SLAVE_ADDRESS_DBG		0x47
+
+#define ARC_RST_VALUE			0x110
+#define ARC_RST_VEC_ADDR		0x0
